p4-16/psa/examples/psa-example-parser-checksum.p4

```diff
@@ -0,0 +1,234 @@
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        buffer.extract(parsed_hdr.ipv4);
+        // TBD: It would be good to enhance this example to
+        // demonstrate checking of IPv4 header checksums for IPv4
+        // headers with options, but this example does not handle such
+        // packets.
+        verify(parsed_hdr.ipv4.ihl == 5, error.UnhandledIPv4Options);
+        // TBD: Update the code below for checking the IPv4 header
+        // checksum with whatever the API we decide upon for a PSA
+        // checksum extern.
+        ck.clear();
+        ck.update({ parsed_hdr.ipv4.version,
+                parsed_hdr.ipv4.ihl,
+                parsed_hdr.ipv4.diffserv,
+                parsed_hdr.ipv4.totalLen,
+                parsed_hdr.ipv4.identification,
+                parsed_hdr.ipv4.flags,
+                parsed_hdr.ipv4.fragOffset,
+                parsed_hdr.ipv4.ttl,
+                parsed_hdr.ipv4.protocol,
+                //parsed_hdr.ipv4.hdrChecksum, // intentionally leave this out
+                parsed_hdr.ipv4.srcAddr,
+                parsed_hdr.ipv4.dstAddr });
+        // The verify statement below will cause the parser to enter
+        // the reject state, and thus terminate parsing immediately,
+        // if the IPv4 header checksum is wrong.  It will also record
+        // the error error.BadIPv4HeaderChecksum, which will be
+        // available in a metadata field in the ingress control block.
+        verify(ck.get() == parsed_hdr.ipv4.hdrChecksum,
+               error.BadIPv4HeaderChecksum);
+        transition select(parsed_hdr.ipv4.protocol) {
+            6: parse_tcp;
+            default: accept;
+        }
+    }
+    state parse_tcp {
+        buffer.extract(parsed_hdr.tcp);
+        transition accept;
+    }
+}
+
+control ingress(inout headers hdr,
+                inout metadata user_meta,
+                PacketReplicationEngine pre,
+                in  psa_ingress_input_metadata_t  istd,
+                out psa_ingress_output_metadata_t ostd)
+{
+    // Table parser_error_count_and_convert below shows one way to
+    // count the number of times each parser error was encountered.
+    // Although it is not used in this example program, it also shows
+    // how to convert the error value into a unique bit vector value
+    // 'error_idx', which can be useful if you wish to put a bit
+    // vector encoding of an error into a packet header, e.g. for a
+    // packet sent to the control CPU.
+
+    DirectCounter<PacketCounter_t>(CounterType_t.packets) parser_error_counts;
+    ErrorIndex_t error_idx;
+
+    action set_error_idx (ErrorIndex_t idx) {
+        error_idx = idx;
+        parser_error_counts.count();
+    }
+    table parser_error_count_and_convert {
+        key = {
+            istd.parser_error : exact;
+        }
+        actions = {
+            set_error_idx;
+        }
+        default_action = set_error_idx(0);
+        const entries = {
```

Comment:
Note that error_idx is not used anywhere anymore, but that's OK -- it serves as an example of how to go from an enum to an integer. :)

Reply:
Agreed, and a comment above this even mentions this explicitly, and explains that it is in the example solely to show an example of how to do this.  If you wanted to put an error index into a header of a packet to the CPU indicating what parser error was encountered, for example.